File: src/compute/moe.ts
/**
 * TS port of the MoE forward pass in training/llmviz_train/moe.py,
 * operation-for-operation, recording the same activation names. Gated by
 * moe.golden.test.ts against PyTorch goldens at 1e-4.
 *
 * A GPT whose dense MLP is replaced by a top-k routed mixture of experts.
 * Attention is reused verbatim from gpt.ts (identical weights). The MoE
 * block scores experts with a router, keeps the top-k per token
 * (renormalized), and combines each expert's dense output by its gate.
 */

import { T } from "./tensor";
import { add, embedding, gelu, layernorm, linear, softmax } from "./ops";
import { getTensor, type Manifest } from "./loader";
import type { Recorder } from "./recorder";
import { dimsFromManifest } from "./util";
import { causalSelfAttention } from "./gpt";

export interface MoeDims {
  n_layer: number;
  d_model: number;
  n_head: number;
  d_expert: number;
  n_experts: number;
  top_k: number;
  vocab_size: number;
  head_dim: number;
  max_seq_len: number;
}

const MOE_DIM_KEYS = [
  "n_layer",
  "d_model",
  "n_head",
  "d_expert",
  "n_experts",
  "top_k",
  "vocab_size",
  "head_dim",
  "max_seq_len",
] as const;

export function moeDimsFrom(manifest: Manifest): MoeDims {
  const dims = dimsFromManifest(manifest, "moe", MOE_DIM_KEYS);
  if (dims.n_head * dims.head_dim !== dims.d_model) {
    throw new Error(`moe dims inconsistent: n_head·head_dim != d_model`);
  }
  return dims;
}

/** One expert MLP: fc → GELU → proj (matches gpt.MLP). */
function expert(weights: Map<string, T>, layer: number, e: number, x: T): T {
  const w = (s: string): T => getTensor(weights, `moes.${layer}.experts.${e}.${s}`);
  return linear(gelu(linear(x, w("fc.weight"), w("fc.bias"))), w("proj.weight"), w("proj.bias"));
}

/**
 * Top-k masked, renormalized gate matrix (T, E) from the softmax router
 * probabilities — gate[t,e] = probs[t,e] / Σ_top-k(probs[t]) when e is in
 * the token's top-k, else 0. Matches torch.topk + scatter in the reference.
 */
function topkGates(probs: T, topK: number): T {
  const [steps, e] = probs.shape;
  const out = new Float32Array(steps * e);
  for (let t = 0; t < steps; t++) {
    const base = t * e;
    // indices of this row's experts sorted by probability, descending
    const idx = Array.from({ length: e }, (_, j) => j).sort(
      (a, b) => probs.data[base + b] - probs.data[base + a],
    );
    let sum = 0;
    for (let r = 0; r < topK; r++) sum += probs.data[base + idx[r]];
    for (let r = 0; r < topK; r++) {
      const j = idx[r];
      out[base + j] = probs.data[base + j] / sum;
    }
  }
  return T.from(out, probs.shape);
}

/** One MoE block (no residual). x already LayerNorm'd: (T, d_model). */
function moeBlock(
  weights: Map<string, T>,
  dims: MoeDims,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const [steps, dM] = x.shape;
  const probs = softmax(linear(x, getTensor(weights, `moes.${layer}.router.weight`)));
  rec?.record(`layer${layer}.moe.router`, probs);
  const gates = topkGates(probs, dims.top_k);
  rec?.record(`layer${layer}.moe.gates`, gates);

  const out = new Float32Array(steps * dM);
  for (let e = 0; e < dims.n_experts; e++) {
    const he = expert(weights, layer, e, x); // dense: every token
    rec?.record(`layer${layer}.moe.expert${e}.out`, he);
    for (let t = 0; t < steps; t++) {
      const g = gates.data[t * dims.n_experts + e];
      if (g === 0) continue;
      for (let c = 0; c < dM; c++) out[t * dM + c] += g * he.data[t * dM + c];
    }
  }
  const combined = T.from(out, x.shape);
  rec?.record(`layer${layer}.moe.out`, combined);
  return combined;
}

/** Full forward: tokens → logits (T, vocab), recording the moe.py names. */
export function moeForward(
  weights: Map<string, T>,
  dims: MoeDims,
  tokens: number[],
  rec?: Recorder,
): T {
  if (tokens.length > dims.max_seq_len) {
    throw new Error(`T=${tokens.length} exceeds max_seq_len=${dims.max_seq_len}`);
  }
  const positions = tokens.map((_, i) => i);
  let x = add(
    embedding(tokens, getTensor(weights, "tok_embedding.weight")),
    embedding(positions, getTensor(weights, "pos_embedding.weight")),
  );
  rec?.record("embed.out", x);

  for (let i = 0; i < dims.n_layer; i++) {
    let normed = layernorm(
      x,
      getTensor(weights, `ln1s.${i}.weight`),
      getTensor(weights, `ln1s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln1.out`, normed);
    x = add(x, causalSelfAttention(weights, dims, i, normed, rec));

    normed = layernorm(
      x,
      getTensor(weights, `ln2s.${i}.weight`),
      getTensor(weights, `ln2s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln2.out`, normed);
    x = add(x, moeBlock(weights, dims, i, normed, rec));
  }

  x = layernorm(x, getTensor(weights, "ln_f.weight"), getTensor(weights, "ln_f.bias"));
  rec?.record("final_norm.out", x);

  const logits = linear(x, getTensor(weights, "lm_head.weight"));
  rec?.record("head.logits", logits);
  return logits;
}

File: training/llmviz_train/moe.py
"""Minimal Mixture-of-Experts (MoE) reference with activation recording.

A GPT whose dense MLP is replaced by a sparse MoE block: a router scores
the experts per token, the top-k are selected and renormalized, and each
expert (an MLP) contributes its output weighted by the gate. Attention is
exactly the GPT CausalSelfAttention (imported, not re-derived).

Computed DENSELY for clarity and exact portability: every expert runs on
every token, then combined with the top-k-masked gate weights — the result
is identical to a gather/scatter sparse kernel but trivial to mirror in TS.

Architecture (pre-norm residual stack):

    tokens -> tok_embedding + pos_embedding
           -> [x = x + attn_i(ln1_i(x)); x = x + moe_i(ln2_i(x))] * n_layer
           -> ln_f -> lm_head (untied) -> logits
"""

import torch
import torch.nn.functional as F
from torch import nn

from llmviz_train.gpt import CausalSelfAttention
from llmviz_train.recording import snapshot


class Expert(nn.Module):
    """One expert MLP: fc -> GELU (exact erf) -> proj."""

    def __init__(self, d_model: int, d_expert: int):
        super().__init__()
        self.fc = nn.Linear(d_model, d_expert, bias=True)
        self.act = nn.GELU()
        self.proj = nn.Linear(d_expert, d_model, bias=True)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.proj(self.act(self.fc(x)))


class MoE(nn.Module):
    """Top-k routed mixture of experts (no residual; the model adds it)."""

    def __init__(self, d_model: int, d_expert: int, n_experts: int, top_k: int):
        super().__init__()
        self.n_experts = n_experts
        self.top_k = top_k
        self.router = nn.Linear(d_model, n_experts, bias=False)
        self.experts = nn.ModuleList(Expert(d_model, d_expert) for _ in range(n_experts))

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) -> (B, T, d_model)."""

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = snapshot(tensor)

        logits = self.router(x)  # (B, T, E)
        probs = F.softmax(logits, dim=-1)
        rec("router", probs[0])

        # top-k select + renormalize; scatter back to a dense (B, T, E) gate.
        topv, topi = probs.topk(self.top_k, dim=-1)  # (B, T, k)
        gates = topv / topv.sum(dim=-1, keepdim=True)
        gate_full = torch.zeros_like(probs).scatter(-1, topi, gates)  # (B, T, E)
        rec("gates", gate_full[0])

        out = torch.zeros_like(x)
        for e, expert in enumerate(self.experts):
            he = expert(x)  # (B, T, d_model), dense
            rec(f"expert{e}.out", he[0])
            out = out + gate_full[..., e : e + 1] * he
        rec("out", out[0])
        return out


class MoEModel(nn.Module):
    """Minimal MoE-GPT language model over a small token vocabulary."""

    def __init__(self, cfg: dict, vocab_size: int, max_seq_len: int = 32):
        super().__init__()
        n_layer = cfg["n_layer"]
        d_model = cfg["d_model"]
        self.max_seq_len = max_seq_len

        self.tok_embedding = nn.Embedding(vocab_size, d_model)
        self.pos_embedding = nn.Embedding(max_seq_len, d_model)
        self.ln1s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.attns = nn.ModuleList(CausalSelfAttention(d_model, cfg["n_head"]) for _ in range(n_layer))
        self.ln2s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.moes = nn.ModuleList(
            MoE(d_model, cfg["d_expert"], cfg["n_experts"], cfg["top_k"]) for _ in range(n_layer)
        )
        self.ln_f = nn.LayerNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size, bias=False)  # untied

    def forward(
        self, tokens: torch.Tensor, record: bool = False
    ) -> torch.Tensor | tuple[torch.Tensor, dict[str, torch.Tensor]]:
        """tokens: (B, T) long -> logits (B, T, vocab)."""
        T = tokens.shape[1]
        assert T <= self.max_seq_len, f"T={T} exceeds max_seq_len={self.max_seq_len}"
        acts: dict[str, torch.Tensor] | None = {} if record else None

        def rec(name: str, tensor: torch.Tensor) -> None:
            if acts is not None:
                acts[name] = snapshot(tensor)

        positions = torch.arange(T, device=tokens.device)
        x = self.tok_embedding(tokens) + self.pos_embedding(positions)
        rec("embed.out", x[0])

        for i in range(len(self.attns)):
            normed = self.ln1s[i](x)
            rec(f"layer{i}.ln1.out", normed[0])
            x = x + self.attns[i](normed, record=acts, prefix=f"layer{i}.attn.")
            normed = self.ln2s[i](x)
            rec(f"layer{i}.ln2.out", normed[0])
            x = x + self.moes[i](normed, record=acts, prefix=f"layer{i}.moe.")

        x = self.ln_f(x)
        rec("final_norm.out", x[0])

        logits = self.lm_head(x)
        rec("head.logits", logits[0])

        if acts is not None:
            return logits, acts
        return logits
